### ISSUE TITLE
Design Direction

### ISSUE THREAD

**Issue by ejelinek** [Mrbrutal/Traincraft-5#2]
So I've gathered from the past that between Mr Brutal and Spitfire, there is a hassle updating the codebase because it is has parts that could have been written better. I understand that, if you worked where I did we do that all day long. So appreciate the position. However being new to MC mod development I'd like a little more help/direction/ideas of where the problems are and why. This would help all of us new to the project understand the where's and why's. On the surface I'd say, update the dependencies and then code till it works. However I get the feeling thats not the direction we want to take. So hence my question/request for help.

**Comment by EternalBlueFlame** [comment]
Seeing as this has gone a month without a reply, I'll pipe in and take a guess towards the point of this one being to design it in a way where it's a lot easier to modify than the previous version, that way when it comes time to move to 1.9/2/etc it will be easier to port than it was previously.
Maybe keep tracks, trains, etc in their own scripts in such a way where they can be modified by a simple call, instead of going
train { all the code for a train}
And repeating all the code for every train/track/etc, but instead making a function so you can just go
train(graphic, sounds[], smoke strength, smoke position, max speed, acceleration, recipe, etc);
and repeat for each.
Probably finding any way to simplify the support and importing of other mods like railcraft and buildcraft in such a way if one of their updates breaks compatibility that fixing it would be simpler, or at least won't break this mod.

This is only a guess and a few very rough ideas towards what he may have meant, I have no clue for sure, but if nothing else this might provide a starter.

Also getting a guide on how to load this into intellij and/or eclipse would be SUPER helpful, I personally can't make hide nor hair of how to do that even after looking into forums for this sort of thing. (Then again the only Java IDE I am really familiar with is Android Studio so it might just be me)

**Comment by ejelinek** [comment]
That is definitely a MUST do, personally I'm still trying to make heads of tails of how the thing is put together, hence my comment around I could just swap out forge/MC for 1.8.9 and keep coding until it all works, but that will be no better than where the project sits now. Maybe I should do that and the heads and tails part will sort themselves out. Once complete there then I would be in a better position to break this thing up into its concerns. I was just trying to avoid that work as it looks to be a ton.

As for setting up intellij, I have managed to get it all up and running. When I get back to the house I'll come up with something and pass it along.

**Comment by EternalBlueFlame** [comment]
Might not be a bad idea, get it working with some notes on what is doing what, before it gets cleaned and organized. More of a mess at the start, but should make the transitions and organization smoother since none of us really know how any of this works yet, just my experience anyway.

And as for the intellij, sweet, that will help a lot. thanks.

**Comment by ejelinek** [comment]
Ok, sorry I was a few days late on this, so I followed these instructions, http://www.minecraftforge.net/forum/index.php/topic,21354.0.html

**Comment by EternalBlueFlame** [comment]
Alright, thanks, i'll give it a shot in a couple days when i get the free time to work on it.

Also something else worth thinking on, the speeds of the various trains aren't great in comparison to other mods, Pixelmon as a great example is a MUCH faster transport than even the highest-end train with nothing being hauled. So I'd think that is another part of the direction to consider, reworking the transport max speeds. I'm not saying make them all super fast, but maybe make the speed difference from train to train a bit more so the higher-end trains can be faster.

**Comment by morphicschris** [comment]
Hey. Excited to see there's some movement on reviving/rewriting Traincraft. It was one of my favourite mods back on 1.6. If you need any help coding I'd like to offer my services. I've got a lot of experience in programming (C# mostly but I did Java back in Uni and the two are pretty close in syntax), but not much in modding Minecraft so will probably need a little help initially finding the right starting point. I've grabbed a copy of the latest code and had a look through it. I'm guessing all the old assets and code aren't in the repository at the moment as I couldn't see anything obvious in the folders that could be them. Let me know if you want my help and where it can best be used. - Chris

Edit - Just seen there is a separate repository for the old code - I'll grab that and have a look through.

**Comment by EternalBlueFlame** [comment]
Yeah, this repository is a blank slate intended for a fresh start, there's really nothing in it yet. The old one has everything but is for 1.6.4 and _supposedly_ is written very sloppy.
Always good to see more hands on deck however.
Honestly I think all 3 of us are really trying to just figure out where exactly to start.
Planning it all out is a good part to focus early, but after that, it's figuring out the code.

One thought might be getting the 1.6.4 version running on 1.7.10, then slowly moving features and functions from it over to this project in a more organized and better designed manner.
Just a thought, I can't say much as I've barley had my head under the hood of this one, let alone the one for 1.6.4.

**Comment by morphicschris** [comment]
Yeah I think getting the existing 1.6.4 version running on 1.7.10 is definitely the best way to start. Once it's running in the later version then refactoring and making improving will be easier. I think trying to fix the old one to work on the new MC version and refactoring the code at the same time will probably cause some headaches while trying to understand what's actually going on in the code.

I've got IDEA running on my computer now and just trying to get forge and the traincraft 1.6.4 project building based on the instructions you posted above.

**Comment by ghost** [comment]
Just wondering if you'll put the mod on curse for download?

**Comment by Fexcraft** [comment]
Eh... just wanted to say that after looking at 1.9 code, most things seems same as in 1.8 (except chat code that was changed).

_edit_ (0059.21.03.16)
after checking it's just needed to rename/reimport classes... then chat works fine, same with  few other things.

**Comment by EternalBlueFlame** [comment]
Putting the mod on curse will have to be under Mr. Brutal's job. Since he has all the rights and whatnot to the mod. And given how busy he probably is....... er...
We could list it in the forums to get it out there, but that's not a direct host, more just a redirect to the github and the main site with some information.

It's good to know about 1.9, that will make things easier, also gives hope for A LOT of other mods.

**Comment by ghost** [comment]
Oh okay. Well, I also wondered if you could port  the 1.6.4 version of traincraft. Just until the 1.7 version is available.

**Comment by morphicschris** [comment]
Hey all. After some tinkering with IntelliJ I've managed to get forge 1.7 to build. Any tips on how I then go about adding the TrainCraft 1.6.4 mod into that? That way I can build it and see what errors will need fixing to get it to work on 1.7.

**Comment by EternalBlueFlame** [comment]
Most likely that will be the entire plan, a port of the 1.6.4 version that will get changed and fixed over time to be said 1.7 version.
A bumpy road no doubt, but I'd wager the easiest and fastest traveled.
Especially since many of us, myself included, have zilch experience with forge's API, or even a good understanding of the way this mod works in the first place.

Importing 1.6.4 on top of forge 1.7, That is the true question.
The source of 1.6 is setup for gradle import, you may be able to use that to import the project on top of what you've already got, since InteliJ is far more kind to gradle than git.
Or you could load them up as two separate projects to study the architecture, learn what to replace in the 1.6.4 variant to get 1.7 forge in there.
I wouldn't say either method is better than the other, more of a 'depends on the person' thing.
There may be other options too, but that's all I got off the top of my head.

**Comment by morphicschris** [comment]
I've managed to import the Traincraft code into IntelliJ now. Building obviously shows a ton of errors, but the first of those seems to be related to missing references to Railcraft. Added the reference to the Railcraft API:

https://github.com/CovertJaguar/Railcraft-API

Build again and I go from 100 errors to... 100 errors. Guess it's limiting the errors it's showing in the window to a maximum of 100! Different errors this time though, so progress. A similar thing applies to IndustrialCraft 2 so I'll import the API for that and see where it gets me.

Most of the errors unrelated to mods seem to be around chat messages, players, and networking. I'm guessing this is the crux of what's been updated between Minecraft/Forge 1.6.4 and 1.7.10. For now I've just got this in a local workspace rather than editing things in Github until I can get it to the point where it's structured correctly and builds (or at least only fails to build because of errors related to code changes that need to be made, rather than just missing references). I'll update if I make any more progress.

**Comment by morphicschris** [comment]
For reference later, here's the IC2 API library (I used the -dev) one.

http://jenkins.ic2.player.to/job/IC2_experimental/

EDIT: Plus the CodeChickenCore and NEI APIs:

http://chickenbones.net/Pages/links.html

Choose 1.7.10 from the dropdown.

**Comment by EternalBlueFlame** [comment]
yeah InteliJ limits error count because of some sort of buffer overflow.

Don't forget buildcraft because I'm sure that will be part of the errors.
[https://github.com/BuildCraft/BuildCraft](https://github.com/BuildCraft/BuildCraft)
[https://github.com/BuildCraft/BuildCraftAPI](https://github.com/BuildCraft/BuildCraftAPI)

It's great to hear you got that far, that's really a significant milestone.
It's likely there will be several hundreds of errors, it is an option to make a branch with the broken code in an attempt to split up the work with everyone. Unless you just wanna blow through it on your own, lol.
If you are going to, be sure to sign the CLA first if you haven't already (its on the main page in the readme).

**Comment by morphicschris** [comment]
Yeah I noticed the Buildcraft references after too. I'd like to branch it and push it back to Git so we can all collaborate on it, but I'll have to have a read about the best way to do so (do I include my entire project structure in the commit or only the traincraft files and leave the user building it to set up the environment?).

I've got some experience with Git and some with Java, but none with the two combined so just need to get my head around the best way to start off with this.

Signed the CLA before I downloaded from Git so we're all good there :)

**Comment by canitzp** [comment]
All of you write a lot about Traincraft and how it can work and now I updated the Repo so everybody can setup it now in 1.9 and Traincraft can get the hole rewrite.

**Comment by EternalBlueFlame** [comment]
Personally, i use Guthub's official program https://desktop.github.com/
Keeps it all generally simple.
There's already a .gitignore which should cover the files that shouldn't be uploaded for you automatically.

canitzp, 5 was supposed to be for 1.7.10 until it was cleaned up enough for easy porting to other versions. Not to mention the code you uploaded doesn't really do anything other than give more of a blank slate than we already had, why Mr. Brutal gave that the 'Okay' i'll never understand. Maybe read the project direction before jumping head first. I do admire your enthusiasm though.

In case anyone needs the pre-1.9 update version, it can be grabbed through the history.
https://github.com/Mrbrutal/Traincraft-5/tree/9c758fa3cf2db75d7bd22afd068e2b15990e9436

**Comment by canitzp** [comment]
Since December I wait for a Workspace I can setup corectly and nobody wants to make a Mod for 1.7 at this time. 1.9 is the newest and a complete rewrite is necessary, so why for 1.7 instead of 1.9. I asked MrBrutal an hour before I began to rewrite it.

**Comment by morphicschris** [comment]
Cheers, I use the desktop one too. Good to know we're on the same page. I'll take a look into the .gitignore and see if I can get something together to push back to Github. Going to be busy doing family stuff now but hopefully should have a little time on Tuesday to give it another go.

**Comment by morphicschris** [comment]
With regards to moving to 1.7 or 1.9, I think the ultimate aim is to be on 1.9, but for now it's an easier step to get it working on 1.7, then work our way up through 1.8 and 1.9. Taking this route means we can make smaller changes and test at points along the way rather than trying to change everything in one go which might be trickier, especially since none of us have a lot of understanding of the overall traincraft code at this time.

There's still valid reasons to maintain compatibility with 1.7 regardless of if we.move to 1.9 anyway, as a lot of mod packs still run on 1.7.10.

**Comment by EternalBlueFlame** [comment]
canitzp, i agree with morph, you say no one wants to make a mod for 1.7, yet here we are, 5 other people here who want to build it for 1.7.
There are more mods available for 1.7 than 1.9 right now, most people will want to go where the mods are. Plus its an easier job to go from 1.6.4 to 1.7 than it is to 1.9.

The entire point of the 1.7 update is to make it easier to port it to other versions and clean up the code. So indirectly making it in 1.7 makes it easier to get out a 1.9 version as well, and hopefully allows us to easily maintain 1.7, 1.8, and 1.9 versions until other mod developers jump on the 1.9 bandwagon and leave us no reason to support the older versions.

Think about this too, the mod is heavily reliant on railcraft, buildcraft, industrialcraft, and a number of other mods that have no 1.9 version yet, even if we did go to 1.9 now, it would be impossible to do it fully because of the missing API. 1.7 however has all those mods and API's.

**Comment by canitzp** [comment]
If Traincraft is ready to use for Modpacks MC 1.7.10 is like 1.5.2 now, still some Packs are running on it but nobody wants to play it. To update a mod from 1.6 to 1.9 it is a lot easier to go straight to 1.9 and not 1.7 - 1.8 - 1.9. Also TC **need** a rewrite and I don't want to rewrite a so complex Mod in 1.7 to port it to 1.8/1.9.

**Comment by EternalBlueFlame** [comment]
I believe we are going to hit an impasse here due to creative differences. So I propose a new option.

Canitzp, you seem to have gotten Mr. brutal's approval for your plans, so why don't you, and whoever else wants to go directly for 1.9 go ahead and work on it from the main branch.

The rest of us who support the 1.7 plan, how about we do so on a branch of the project, aimed at that direction.

this way everyone gets what they want.

**Comment by canitzp** [comment]
It's a plan but it is good to have two different source codes? If the source code don't have so many differents the plan is a good idea, but otherwise it can confuse people

**Comment by EternalBlueFlame** [comment]
Well I don't see another option.

A number of us want it for 1.7 because there are basically no other mods available for 1.9, and having only traincraft is kind of boring, not to mention the extra difficulty with the missing API from said other mods, which this one relied on a good bit.
But apparently there's want for that outcome as well, and Mr. Brutal has given it the okay on the main branch.

So the only answer I see is to have separate branches, the main one marked for 1.9 and a secondary branch marked for 1.7.10(later to be multi-version).

This is a common thing that happens in open sourced projects when creative differences hit.

My personal server runs 1.7.10 currently, due to a lack of mods on other versions, so even if I'm alone in it, for my own reasons I have to support the 1.7 move, and I'm sure there are others in the same boat with me.
Just as I'm sure there are others in the same boat with what you say.
Although I can't speak for anyone here other than myself, so we will have to see who whats to go where.

**Comment by morphicschris** [comment]
I'm in the 1.7 camp too, but I'd like it if we can find a way to keep all versions from the same source code. Without looking into this for Java I can't say whether the following is possible or not, but the way I'd handle this in a C# project would be to have one solutions (i.e. master project file) which contains separate projects for each version, which share the same source code files but use pre-processor directives (hash-defines) to only include the relevant code bits for each version. So for example it might be (and I know this code is nonsense, but for the sake of argument):

```
public void SomeMethod()
{
    SomeOtherMethodCall();

#if VERSION17
    Call17Code();
#elsif VERSION18
    Call18Code();
#elsif VERSION19
    Call19Code();
#endif
}
```

Or something to that effect. Hope that makes sense. For static files like XML, you just reference the file relevant to the version you're working with, so have separate files for 1.7, 1.8, and 1.9.

**Comment by canitzp** [comment]
This don't wok with Minecraft, because the are too many differents between Versions. This only can work with a little update like 1.7.2 and 1.7.10

**Comment by morphicschris** [comment]
So perhaps the way to go then is to branch the code as eternalblueflame has suggested, but do so for each version? The code common to all versions is stored in the main branch, which is then pulled and merged into the version specific branches. That way we keep the core of things common but only specialise the parts needed for each version. The base point of this would probably be from the existing 1.6.4 version.

**Comment by EternalBlueFlame** [comment]
I have to agree with canitzp this time, I didn't think about it until now, but maintaining all the versions in the same repository would get complicated for new people, unless forge offers a back-porting API like Android where you can call something along this line

> @TargetApi(1.7.10)
>     public void doAThing(){
>         //do a thing
>     }
> @TargetApi(1.8)
>     public void doAThing(){
>         //do a thing
>     }
> @TargetApi(1.9)
>     public void doAThing(){
>         //do a thing
>     }

In C# terms:
The problem is that in java its a dependency thing, imagine if each version of forge is a different version of .net framework, you can't even compile a .net 2 program when you have code, used or not, lingering from .net 4.

The only solution to this issue without a back-porting API is to use different scripts which have to manually be changed.

Long run I would agree it is more simple to do a different branch for every version.

Although, if you look at the rest of the  modding scene, very few wanted to make mods for 1.8, the buildcraft team, cofh, and the people behind millénaire  have said they would rather wait for 1.9 then bother with 1.8, this was also assisted with a mutual distrust across the community because of microsoft owning minecraft as of 1.8. Which seems to have cleared up as of 1.9.

So long run there may only be the 1.7 and the 1.9 versions due to a lack of mods/interest in 1.8. Which is still only 2 branches, not counting the 1.6.4 branch(which I personally see no point in maintaining), but that's just my opinion/thoughts.

//EDIT
Also on the point of sharing code between branches, the only point in that would be saving storage space, which isn't a problem. Mostly it just makes things more complicated from having to be constantly pulling this and that from branch to branch.
Plus with the different design directions, not even counting forge incompatibilities, the code is likely to be VERY different between the two.
I mean the 1.7 plans were to take the 1.6 code and fix it up after getting a working port, while as the 1.9 plans are pure from scratch re-write.

**Comment by morphicschris** [comment]
OK, sounds like having completely separate branches is the way to go forward then. How do we proceed with this? Do we need MrBrutal to set up a 1.7 branch?

**Comment by EternalBlueFlame** [comment]
Well my reply is way late, sorry.
I actually asked MrBrutal this directly a good while ago, before the open source code was available.
Given that it's not really part of the official branch, but rather a spinoff of it, we won't need MrBrutal to set up the 1.7 branch, one of us would just have to fork the main branch, and make sure its noted on the fork to be for a 1.7 version that isn't intended to be merged with the main branch.

The fork should also allow for bug reports, pull requests, wiki, etc.. Totally separate from the main branch, which is convenient.
I'll set that up now, just make it a little easier on everyone.
[https://github.com/EternalBlueFlame/Traincraft-5/tree/master](https://github.com/EternalBlueFlame/Traincraft-5/tree/master)
If you'd rather be the one managing it just let me know, I can transfer it to you. Although I doubt it makes much difference either way.

//edit
I also added you as a collaborator morphicschris. So you shouldn't need approval to push a commit to it. I think.

**Comment by morphicschris** [comment]
Awesome, thanks for adding me. I'll grab a new copy on my desktop to start making changes on. I'm happy with you managing it if you want to.

**Comment by EternalBlueFlame** [comment]
Alright, Just a heads up on that, the fork has the most current 1.9 code (I cant fork from an earlier version), so we will probably have to overwrite/remove a bunch of stuff.
Any further conversation on this should probably be moved over to the project, no need to clutter things here, haha.
[https://github.com/EternalBlueFlame/Traincraft-5/issues/1](https://github.com/EternalBlueFlame/Traincraft-5/issues/1)

**Comment by NitroxydeX** [comment]
Didn't read anything here but this got discussed on the Forum for ages. The best thing is to port this to 1.9 and than backport the rewrite to 1.7.10 (which is the easier way) Think about this: If you do a rewrite for 1.7.10 everything needs to be redone in 1.8 and 1.9 (1.8 changed a lot and 1.9 changed a lot aswell)

**Comment by EternalBlueFlame** [comment]
Sorry, this was a long conversation, should have left a cliffnotes version.

Essentially we have all agreed that we have our own reasons for perusing it under 1.7 or 1.9 so there are two separate versions;
the main branch here is dedicated to the 1.9 rewrite, remember that this mod was HEAVILY dependent on the railcraft API, and should still even in 1.9 unless you want to make your own cart and fluid handlers, however RailCraft/BuildCraft/IndustrialCraft/GalactiCraft/other-machine-cart-fluid-mods-here, are not fully in 1.9 itself, so good luck there.

meanwhile there is also the 1.7 port which I have linked in my previous reply, that version is down to a few hundred errors vs the couple thousand it started with and the related mods have stable and well known API's. This version will probably move to 1.9 when other mods have as well.

Now when it comes to porting I have had to do some extensive research on it, the vast majority of the changes between 1.7 and 1.8/1.9 is that the imports have different addresses now, very tedious work, but very easy fixes, there are other minor changes but most of it is the imports.
The change from 1.6.4 to 1.7 and later is the part where it's actually difficult, and these same difficulties lie in 1.7 and 1.9 either way as long as you are coming from 1.6.4.

But as I have stated, we all have our own reasons for perusing it under whatever version. You of course are no exception, and I fully respect that.
Both projects are open, and we'd be glad to have your help on either/both.

//edit
To further expand it now that I have more time to post (the above was rushed).
the major changes were that in 1.7 we moved from ID's to blocks, and there was a new networking system, configs, villages, and various misc variables were also changed.
Nothing in regards to that is changed in later versions, so it has to be done in 1.7/8/9 regardless.
In 1.8 really the only things changed were in relation to the imports, you don't get fixes easier than that, you can probably just regex those fixes in notepad++. But that has to be dealt with in 1.8 and 1.9.
then 1.9 comes, now I have not gotten very far into this part, but most devs are very reassuring that porting a mod from 1.8 to 1.9 is very minimal effort, and I honestly doubt anything that had to be changed to go from 1.6.4 to 1.7/1.8 was changed again.

So in reality, going from one version to the next actually does make it easier and allows you to get the same job done, for really no extra effort, and cover the areas that are in demand.
A lot of people want to make mods for 1.9 granted, but there are almost no mods for 1.9 ergo, a small actual playerbase.

And by the time its working for 1.7 then 1.9 should have a good number of mods and playerbase so we can just skip right over 1.8 and go for 1.9.
Of course as stated above the 1.8 changes still have to be made even when skipping it because said changes exist in 1.9.

**Comment by canitzp** [comment]
The rewrite Traincraft without any API and one big change from 1.7 to 1.8 are the big render changes and BlockProperties, so all rendering stuff has to rewrite there. I develope the rewrite in 1.9, cause when Traincraft is at a state to open it for the public, 1.9 is the normal modding version. Also a lot of mods (look at curseforge) are updated to 1.9

**Comment by NitroxydeX** [comment]
Wait.. To get you right here. You rewrite Traincraft without an Traincraft API?

**Comment by EternalBlueFlame** [comment]
I haven't seen anything in relation to render changes, I guess that's a hurdle to jump over when I hit it.
But my main point was in a port, even if you go directly to 1.9 from 1.6, the changes necessary for ports to 1.7 and 1.8 have to be done the same either way. So it's not any extra effort to pop out a 1.7/1.8 version along the way.

What I should have clarified more, and I apologize for not doing so, is that since the 1.9 version (here) is a complete rewrite, that really won't matter since it all has to be done from scratch anyway, unlike in the 1.7+ port I linked previously.

As far as mods being updated to 1.9, does 1.8.9 count as 1.9? Because most of the ones I mentioned, and most of the mods anyone actually uses for that matter, are only to 1.8.9. (a couple are still back at 1.7.10)

**Comment by gioporta** [comment]
A lot of people and modpacks are still on 1.7.10, so I agree with what EternalBlueFlame is doing. I will admit the 1.8 and 1.9 modding scene has been increasing very quickly in the past 2 months, however major mods at 1.7 are what are holding me, as well as others, back. 

Updating in chronological order (1.7, then 1.8, followed by 1.9) is the logical thing to do in my opinion.

**Comment by EternalBlueFlame** [comment]
Oh NitroxydeX, late reply, sorry.
I think what canitzp was trying to say was the plan for 1.9 is to rebuild without the dependency API, the old traincraft was fully dependent on the Railcraft, Buildcraft and ChickenCode API's to function.

Of course this means making fluid managers, GUI, minecart Handlers, etc, will have to be made from scratch.
I really don't blame him for this, because it's annoying having to rely on so many dependencies, and once the 1.7 one I'm working on is actually working, I'd like to do the same, at least to get rid of the Chicken Code dependency.

As for GiovanniPorta, this is why I split the project into the 1.7 port, when it comes to a full rewrite, 1.9 directly is the way to go, but porting, chronologically is the easiest option, with the best results.
But we all have our reasons for working on one or the other, so this gives everyone the choice of what they want to do.

**Comment by NitroxydeX** [comment]
Okay, but I hope canitzp is building Traincraft with an official Traincraft API, like BiomesOPlenty does the same, so it's easier later to add things to Traincraft - even for Traincraft Addons.

Dependencies are good - but only if you are the maintainer of the other Mod. ;)

Railcraft is a nice thing but only for Minecarts. We need some really cool Tracks in Traincraft. AlexIIL started a really good idea over here: AlexIIL/TrainCraft-1.8.9 we should think about this. You see some of the 3D-Tracks here: AlexIIL/TrainCraft-Ideas#2, here AlexIIL/TrainCraft-1.8.9#6 and here AlexIIL/TrainCraft-1.8.9#13. His Idea was that the Rails connect to each other if you place them on top of each other. He mentioned this here AlexIIL/TrainCraft-1.8.9#15

About 1.8.9 and 1.9 - 1.8.9 belongs to the 1.8-branch not the the 1.9-branch. Forge changed something in the GameRegistry and many things got renamed in 1.9.

The problem with the 1.8-branch was that it got hated without any reason. People hated it and developers had no motivation to port their mods to this branch. Buildcraft stopped at 1.7.10 because there were problems with the obj-Render that got re implemented later in the 1.8-branch. This were things to deal with at the beginning and made it even harder to get famous mods to 1.8.
Some time ago many developers started porting their mods over to 1.8 because they prepared theirselves for Minecraft 1.9. We should skip the 1.8-branch it would not be the best Idea to have a mod for an unloved Version that is not really supported by other developers.

If there is a 1.7.10-port of Traincraft it's good - for those ones who are still on 1.7.10. But how long do you need for it? Do you think you finish it before 1.9 explodes with famous mods? We shouldn't work on too many 1.7.10 ports, I saw a lot of them it's making me sick why not work together? If the port can't be stable long before 1.9 development explodes we all should concentrate on 1.9. But thats just my 2 cents on this topic ^^ 

But we should think about making complete new models for Traincraft 1.9!

**Comment by canitzp** [comment]
Yes I add a api to Traincraft and there is also one thing in the api folder. https://github.com/Mrbrutal/Traincraft-5/tree/master/src/main/java/si/meansoft/traincraft/api

**Comment by NitroxydeX** [comment]
You should think about issue labels here. So we can seperate bug reports from ideas. It's better than a forum.

**Comment by EternalBlueFlame** [comment]
Sorry, day late, internet was out.

Railcraft has its own set of, sort of, different rails with vast features, it also has a fluid manager, like for steam boilers, plus a custom cart system.
The old version of Traincraft used Railcraft's API for the fluids, like steam boilers, it also used the custom carts.
Since it was only the API of Railcraft, we didn't get any of the rails in Traincraft, but if you had both mods, you can use them.

The fears of 1.8 were partially justified, since that was the version right after mojang was bought by microsoft. And everyone loves to hate microsoft. Even Spitfire commented he was leaving partially because he feared Microsoft would destroy Minecraft, there were other reasons for him too, but still.
Needless to say, I think everyone was relieved that those fears were nullified.

The 1.7 branch is down to around 200 errors before debugging can commence, I'm really the only one working on it right now, the other guy helping has been very busy, so I can't give a timeframe, but numbers wise, its not far from working. More help would naturally speed it up.
You can take a look at it if you want: [https://github.com/EternalBlueFlame/Traincraft-5](https://github.com/EternalBlueFlame/Traincraft-5)
But you have to remember, that branch is a port of the 1.6.4 version to 1.7.10, this 1.9 branch here is a complete rebuild.
A port is faster and easier (good for noobs to the MC modding scene like me), but a rebuild will have (potentially) better performance, and overall internal design, long-run.
The 1.7 version may come to 1.9 as well, but generally theres a good number of reasons for splitting the work.
For me personally, due to my experience in java/C# and lack of experience in MC modding, I can get a LOT more work done in a port, than I can a rebuild. It also gives me a chance to learn what I'm doing.

As for skipping 1.8, I agree to skip it, just to ignore having to make builds for it, but at the same time, any changes that have to be made for 1.8 still have to happen for it to be made for 1.9.
It's not like the update from 1.8 to 1.9 obsoletes EVERYTHING, similar how the update from 1.6 to 1.7 doesn't, only a few features here and there get changed, like ticks, ID's, and networking.

Also, on the labels, the labels are on a basis of a per-issue thread (Github limitation), so any requests, issues, suggestions, ideas, etc, that would merit a Label should be mentioned in a new issue thread. This specific issue thread is more of a general discussion. Or at least, that's what it has become.
The same goes for the 1.7 port.

//edit
if we REALLY do want to go for different rails and new graphics, I'd say going for something more like Rails of War and RealTrains.
[http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/wip-mods/1443999-1-7-10-rails-of-war-under-reconstruction](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/wip-mods/1443999-1-7-10-rails-of-war-under-reconstruction)
But with that, we lose all the features of minecarts and rails from Minecraft and Railcraft, unless we remake them ourselves, which would be a lot of work for very little gain.
Honestly the best option in my opinion is to extend on what Railcraft and minecraft have for rails, to add some smoother turns. Sort of similar to what you showed. That's a lot less work and the same general result.

**Comment by Spitfire4466** [comment]
Glad to see the mod is being actively worked on and design direction being discussed, keep up the good work!

**Comment by StargateMC** [comment]
As a server hoster - my recommendation would be to upgrade to 1.7.10 and Then to 1.9 when sponge is available. There is little benefit running 1.8.

A lot of packs rely on Thermos or Cauldron derivatives to provide proper grief prevention, economy etc.... Which runs on 1.7.10 and will not be made on 1.8.

Once sponge is ready (and has a decent plugin base), most of us will be moving to 1.9+ and utilising that as a replacement for plugins.

In addition to this, many major mods do not yet support 1.9 or 1.8.x... and have stuck on 1.7.10 for this reason.

Just a bit of feedback.

**Comment by EternalBlueFlame** [comment]
@DraksterAU Thanks for the feedback, due to the desires of various contributors, we have already split the project into a 1.9 rewrite, and a 1.7.10 port.
This project is for the 1.9 rewrite.
For the 1.7.10 port see here:
[https://github.com/EternalBlueFlame/Traincraft-5](https://github.com/EternalBlueFlame/Traincraft-5)

Remember that the port, even when it's working, will be very inefficient, mostly due to the design of the 1.6.4 version. (Thus the point of the rewrite, which we leave to the more experienced contributors, aka not me.)

Another thing is that, from what I've heard, sponge was intended to be, mostly, forge compatible, similar to Cauldron, so if the 1.9 doesn't work on sponge when it's ready, adding support for it _shouldn't_ be a huge problem. Well assuming all goes to plan on the sponge side.

**Comment by StargateMC** [comment]
Sponge works with forge. You won't need to worry about support.

The point is it has like 10 usable plugins whereas 1.7 has tens of  thousands.

Really would have gone 1.6>1.7>1.9 if I was you guys... Making gradual changes rather than making the mod unusable for months or years.

All good though... Lol!

Speaking as a large server hosted here.

**Comment by EternalBlueFlame** [comment]
Good to know I heard right then.

I partially agree to the 1.6>1.7>1.9 part, but you have to remember too that in a port, going from version to version is easier and you kinda gotta do it that way anyway, but a rewrite is easier if you just do it in the newest version, so then you don't have to port the mod you just built from scratch(at least not for a while).

The 1.9 version wouldn't even really be possible right now if it wasn't a rewrite from scratch, because what used to be necessary dependencies like railcraft, in the old traincraft, aren't really settled into 1.8 yet, let alone 1.9.

**Comment by StargateMC** [comment]
Lol you seem more focused on a rewrite than actually getting things working.

There's a reason rail craft haven't updated. It's not worth it.

**Comment by canitzp** [comment]
@DraksterAU Please look at all the updated Mods at curseforge. Every day new mods are updated and there are a lot of modpacks for 1.9. I rewrite Traincraft for 1.9 and I need month to get it to a good state, cause nobody helps me with the rewrite and I would appreciate it and Traincraft isn't the only mod I develope.

**Comment by NitroxydeX** [comment]
@DraksterAU ehm... I think you never read anything anywhere. The change from 1.7.10 to 1.8 was so massive that many developers just waited to get a really good forge experience on 1.8. Many big mods also rely on the obj-render which was removed with 1.8 and later added back in as a complete rewrite but it wasn't finished at the start.

I know that Railcraft is already working on a port so: This is no valid argument. The thing about the 1.9 rewrite is that if it's finished 1.9 will be the major mod version. It's a MASSIVE rewrite of a really BIG mod. So don't expect this too soon. 1.7.10 port will be there in a short.

But also Sponge has some problems. It can't interact with all the things from forge because some of the part-devs of Sponge refuse to contribute to their parts. So e.g. the new land mass generator in sponge isn't working with other world-types because the dev just said nope. Most things are easy to maintain in Forge itselfs. So if you are a server owner hire a developer to get your own personal security-plugins for Forge. Thats works the best - not through an API like Sponge.

**Comment by NitroxydeX** [comment]
@canitzp maybe you should think about publishing your rewrite into an own repo. It's easier for other developers to maintain it with you. If you just push your things outta nowhere not all developers know if you haven't done already more than in this repo.

**Comment by StargateMC** [comment]
@nitroxydeX  I'm saying move from 1.6-1.7-1.8-1.9 and by the time you support 1.9 most of the big mods including rail craft will too.

Sponge world generators work fine, your mistaken.

The main issue is not all developers are willing to or willing to let others port their Bukkit code to it yet. 

Sponge is built by forge, cauldron and some Bukkit devs - do not think of it as an Api, more an extension of forge itself.

You'll find most larger server owners do not use vanilla forge. There is no way to properly and easily maintain a half decent economy, or most other server side logic plugins that are typically provided from bukkit  (such as grief prevention). It's the whole reason why Thermos or Cauldron Exist.

You would also realise that most larger servers run this if they are not a basic survival or creative experience.

Finally - not all of us have spare cash lying around. Some of us host a network out of our pockets.

I've unsubscribed to this thread. Good luck with it all.

**Comment by NitroxydeX** [comment]
@DraksterAU But you don't seem to get the things out of a developer position. The rewrite needs a massive amount of time. If we would start it in 1.7.10 there would be a 1.7.10 version in approximately 6 months? In this time most big mods changed to 1.9 - Traincraft would still stuck at 1.7.10 then. 

Then you need to make big changes for the 1.8.9-port and after that you need to make big changes for the 1.9 port. This could take up to 12 months and what is the result? Little Kids crying for Traincraft (again).

Buildcraft is actively developed on the 1.9-port, BiomesOPlenty has already a 1.9-port and other big mods are WIP Alpha for it. And you have to be fair: If big mods port from 1.7.10 they will leave out the 1.8-branch because they already did the last whole year.

**Comment by EternalBlueFlame** [comment]
so much to reply to...
@DraksterAU 
we are equally split between rewrite and making a working port, as i said we are all just contributers with our own ideas where things should go, thats why we split into two groups.
if you wanna see a lot of effort on the rewrite, look here. by the time 1.9 is widley accepted, this branch should be working, or close.
if you wanna see a lot of effort on actually getting it working, look to my 1.7 branch, which will be ready significantly sooner.
[https://github.com/EternalBlueFlame/Traincraft-5](https://github.com/EternalBlueFlame/Traincraft-5)

A vanilla forge mod for managing econemy is millenaire. I prsonally use that and it's great for econemy in more ways than any bukkit plugin. theres also a number of other bukket-esk tools for forge.
also the primary reason cauldron exists is because forge is inefficient, i use pure forge mods but on cauldron because it runs so much better than forge.

One more thing, thanks for being passive agressive then ragequitting like a whiny brat after ignoring half the things we say, please dont come back.

@NitroxydeX 
as for publishing the rewrite to its own repo, this is said repo for that. we've agreed this will be for a 1.9 rewrite, since this branch was always intended for a rewrite, but i made my branch, linked earlier in this text, for a port. So canitzp is free to continue the work here, which I may add, he seems to be doing very well at.

**Comment by EternalBlueFlame** [comment]
one more thing @DraksterAU dont pull out the server hosting fees like you are paying us to write the mod by paying your electric bill. I personally do _NOT_ appreciate that sort of comment.
again, dont come back to this chat.

**Comment by EternalBlueFlame** [comment]
oh actually @canitzp I have a legitimate request I wish to make, since you are a contributor to this branch.
I would like you to make an issue thread that starts by listing this projects goals, what it is, and notes the 1.7 port branch for the impatient people.
Along with an actual general discussion thread.
I did something very similar in the 1.7 branch, see here:
[https://github.com/EternalBlueFlame/Traincraft-5/issues](https://github.com/EternalBlueFlame/Traincraft-5/issues)
If you would like some help in doing that, I'd be glad to assist.
I ask this because it has been seeming lately that this, what feels like a forum, causes more confusion to new people, rather than actually being helpful.

I also opened a discord channel for traincraft, if you approve of that idea you can list it too. (its kinda like IRC, but more feature-rich, even works in browser)
https://discord.gg/0zzHVjy4UIqVJskM
This link defaults to the 1.9 general chat. There are channels, and permissions, setup for general and dev channels, theres also roles for devs, admins, etc.

**Comment by NitroxydeX** [comment]
@EternalBlueFlame I don't think you got me right here. It's just he pushes the things out of an invisible repo instead of cloning this one over to his PC. So: Thee other developers don't know if he made a change on his personal end and just didn't pushed it. Makes it harder after that to merge the things together.

If he just don't want to push to much commits to this repo and just push a big change-packet it would be a better way to have an own repo to it. But let's see what he will reply. :)

If you are done with your 1.7.10-port it should be merged over here in it's own branch. This is much easier to maintain for all people. 

And this Repo should be renamed. Remove this not necessary "-5" @Mrbrutal

**Comment by EternalBlueFlame** [comment]
Okay I understand now, thanks for clarifying, and I do see your point.
Although it is possible he uploads the changes as he makes them (like I do), so doing that would be rather pointless.
But we'll have to see what he has to say on the matter.

The 1.7.10 port is not done(although its not far, mostly the networking is stopping it, and I need some help with that), but it is in it's own branch alongside branch already, unless I have done something wrong, or unless you mean to merge the two branches, which would break both mods?

**Comment by NitroxydeX** [comment]
Just merge it later over to this official repo. I think this repo will get a jenkins build server. So the master is always the newest version and we have also the 1.7.10-version in one place.

**Comment by EternalBlueFlame** [comment]
I'm very curious, if I understand you correctly, you're saying to merge the two branches, at some point, how do you suggest we do it where it doesn't break both mods?
Because a merge would cause overwrite/conflict with several key files in both mods.
Gradle files as an example.
Please correct me if I'm wrong.

**Comment by NitroxydeX** [comment]
Every branch has it's own file-system. This wouldn't conflict. Look at other big mods. They do the same.

**Comment by EternalBlueFlame** [comment]
Okay, now I get it.
And I'm guessing a fork is different from a branch?
(There's already a fork setup).
Sorry for all this confusion.

**Comment by NitroxydeX** [comment]
A fork is a copy of the repo in your own account. Only you can edit it. The master only changes if you update it from the original repo.

**Comment by EternalBlueFlame** [comment]
Interesting...
A branch sounds like something a contributor or the owner of this branch will have to setup then, for the 1.7 port to move to. correct?
Also forks can be set to be open for editing, as I've done. But that is disabled by default.

**Comment by NitroxydeX** [comment]
You did the same on your repo. 1.7 is the branch.

**Comment by canitzp** [comment]
@NitroxydeX I commit the 1.9 Traincraft rewrite to the master branch of this repository. If @EternalBlueFlame want to push their 1.9 port into this repository he use the 1.7.10 branch. If 1.10 Forge has a recommend state before I can have a state of Traincraft to release it, I create a new branch 1.9 and at the master branch the 1.10 developing is active. Why I should use a different repository or  fork of this one, if I can push direct on this without a pull request?

**Comment by NitroxydeX** [comment]
If you always push your changes its okay. :)

**Comment by MggMuggins** [comment]
I feel like I should have been paying more attention to this much sooner, then I could have tried to push things in a possibly faster direction. Now, I have practically no java/programming experience in general, but I have enough of a clue to kind of understand what is going on. 

I'll say it plainly, hoping that I'm not offending anyone. I think it might have been much easier to get more work done more quickly and efficiently if this had been handled differently. According to **@EternalBlueFlame**, most of the work for a 1.9 port/rewrite would have had to be done for a 1.7 one as well. Even though I know that this isn't the way that this project has gone, it seems to me that if all of the contributers to this project had done a 1.7 re-write, then ports to 1.8/1.9 could have been completed in a more timely manner, although the 1.7 re-write would have taken longer. You might have found more success pooling all of your resources, and even though you may not want to work on a 1.7 re-write, it would make the project go faster, in addition to making the problem of having two completely different mods with differently formatted code. I will admit that I think that a 1.7.10 version is as important if not more so, than a 1.9 version, however, I recognize that some people would in fact use it, and therefore, is something that would probably be worth some time. Just my two cents.
~Samwise

**Comment by EternalBlueFlame** [comment]
@MggMuggins the 1.7 port has been in the works for a long time now (since may). And is nearly complete aside from some misc crashes and networking issues.
Take a look at: [https://github.com/EternalBlueFlame/Traincraft-5/](https://github.com/EternalBlueFlame/Traincraft-5/)

After a port is complete then we clean it up and look into moving to newer minecraft when there is demand for 1.9.4 since after cleanup, porting it again won't be anywhere near as difficult.

It would have been faster and more efficient to do a rewrite **only** if the team was previously experienced, which we weren't. We are now, but now it's kinda late.
Not to mention, many of us (myself included) did not want to do all the work for a rewrite and get basically no rights on the project for all the work done. At least with a port, end of the day, majority of the effort was still the original developers.

Also something to keep in mind is that with a rewrite, keeping backwards compatibility is VERY difficult, if not impossible given how messy TC code is/was.

One more thing, way to necro an ancient thread topic.

**Comment by canitzp** [comment]
To discuss we now have the discord chat: https://discord.gg/0zzHVjy4UIqVJskM
